### PR TITLE
build: Drop make dist in gitian builds

### LINKED ISSUE #16588: Quick version string hack that was never fixed
In #11097 we added a quick version string hack that seems to do the following:

-----

When Gitian building, ***after*** generating the source tarball (a la `make dist`), we invoke `make -C src obj/build.h`, as `make dist` does not: https://github.com/bitcoin/bitcoin/blob/b499d8576f30d13dfe9000852afe47ee1c28cfbe/src/Makefile.am#L245-L248
We can see that this calls `share/genbuild.sh`, which generates a `src/obj/build.h` containing version information obtained from `git`: https://github.com/bitcoin/bitcoin/blob/b499d8576f30d13dfe9000852afe47ee1c28cfbe/share/genbuild.sh#L1-L53

Then, for each architecture we build, we neuter `share/genbuild.sh` (as normal `make` will try to regenerate `obj/build.h`, but now lacks the build information as we're not in a git repository anymore) and copy over the generated `src/obj/build.h` mentioned above. All of this is being done before `configure`ing, `make`ing, and `make install`ing for the architecture: https://github.com/bitcoin/bitcoin/blob/b499d8576f30d13dfe9000852afe47ee1c28cfbe/contrib/gitian-descriptors/gitian-linux.yml#L168-L177

-----

Perhaps what we can do is simply add `src/obj/build.h` to `EXTRA_DIST`, and generate it in a `dist-hook`? Do non-existent files in `EXTRA_DIST` get their targets called? Source: https://www.gnu.org/software/automake/manual/html_node/The-dist-Hook.html

@theuni @luke-jr would love some more context here.

### LINKED ISSUE #18547: Followup for #18331 "build: Use git archive as source tarball"
After the merge of #18331, there are now two ways to create source tarballs.

1. `make dist`
2. `git archive`

It would seem that both methods are being used in our Gitian descriptor.

First, we run `make dist` to create a bootstrapped tarball: https://github.com/bitcoin/bitcoin/blob/c5966a87d1fdd7a98f2baee5b2deddd541fdfb5a/contrib/gitian-descriptors/gitian-linux.yml#L143-L148

Which is used by each host to build the binary tarballs in the for loop:
https://github.com/bitcoin/bitcoin/blob/c5966a87d1fdd7a98f2baee5b2deddd541fdfb5a/contrib/gitian-descriptors/gitian-linux.yml#L154-L191

Then, we use `git archive` to create the final source tarball: https://github.com/bitcoin/bitcoin/blob/c5966a87d1fdd7a98f2baee5b2deddd541fdfb5a/contrib/gitian-descriptors/gitian-linux.yml#L194

It seems to me that perhaps it'd be good to reinstate the behaviour whereby our binary tarballs are built from the source tarball, just for robustness's sake.

A simple fix might be:
1.  switch the earlier `make dist` to `git archive`
2. `./autogen.sh` for each host
3. Remove the last call to `git archive`

Perhaps this can even allow us to specify the `no-dist` automake option and clean up `dist` related targets in `Makefile.am`.

### PR DESCRIPTION
**Pull request by hebasto**
After the merge of #18331, the packaged source tarball is created by `git archive`, but the binaries are built from another one which is made by `make dist`.

With this PR the only source tarball, created by `git archive`, is used both for binaries building and for packaging to users.

Close #16588.
Close #18547.

As a good side-effect, #18349 becomes redundant.

**Change in behavior**

The following variables https://github.com/bitcoin/bitcoin/blob/1b151e3ffce7c1a2ee46bf280cc1d96775d1f91e/configure.ac#L2-L6

are no longer used for naming of directories and tarballs.

Instead of them the gitian descriptors use a git tag (if available) or a commit hash.

---

Also a small refactor commit picked from #18404.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Open-Close to re-run ci. See https://github.com/bitcoin/bitcoin/issues/15847#issuecomment-487591130

**Comment by MarcoFalke** [comment]
Travis run here: https://travis-ci.org/github/bitcoin/bitcoin/builds/672303495 . Not sure why it is not linked.

**Comment by laanwj** [comment]
Concept ACK

**Comment by laanwj** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
Looks like `bitcoin-0.19.99-win64-setup-unsigned.exe` is named incorrectly

**Comment by hebasto** [comment]
Updated 8d0c980e6425012f772c89ff4b8346ba3f459151  -> 19a51a83454b39698b76651a40e5ebf31c61b234 ([pr18556.01](https://github.com/hebasto/bitcoin/commits/pr18556.01) -> [pr18556.02](https://github.com/hebasto/bitcoin/commits/pr18556.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18556.01..pr18556.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18556#issuecomment-611689873)
> Looks like `bitcoin-0.19.99-win64-setup-unsigned.exe` is named incorrectly

File: contrib/gitian-descriptors/make_git_archive [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#
+# A helper script to be sourced into the gitian descriptors
+
+mkdir -p ${OUTDIR}/src
+RECENT_TAG=$(git describe --abbrev=0 HEAD)
+if [ $RECENT_TAG = $(git describe HEAD) ]; then
+  if [[ $RECENT_TAG == v* ]]; then
+    VERSION=${RECENT_TAG:1}
+  else
+    VERSION=$RECENT_TAG
```
Comment (MarcoFalke) [inline contrib/gitian-descriptors/make_git_archive:13]: what is this condition for?

**Comment by MarcoFalke** [comment]
ACK, can squash the first two commits?

File: contrib/gitian-descriptors/make_git_archive (hunk 1 above)
Comment (hebasto) [reply contrib/gitian-descriptors/make_git_archive:13]: We tag releases with the prepended "v", like "v0.20.0rc1".

This `if` statement just removes it on purpose to include the version string (e.g., "0.20.0rc1") into file names.

Comment (MarcoFalke) [reply contrib/gitian-descriptors/make_git_archive:13]: Oh, I meant the else branch

**Comment by hebasto** [comment]
Updated 19a51a83454b39698b76651a40e5ebf31c61b234  -> 2aa48edec0101f8a77a2189244fc62722ff7a123 ([pr18556.02](https://github.com/hebasto/bitcoin/commits/pr18556.02) -> [pr18556.03](https://github.com/hebasto/bitcoin/commits/pr18556.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18556.02..pr18556.03)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18556#issuecomment-615228781)
> ACK, can squash the first two commits?

File: contrib/gitian-descriptors/make_git_archive (hunk 1 above)
Comment (hebasto) [reply contrib/gitian-descriptors/make_git_archive:13]:
```
if [ $RECENT_TAG = $(git describe HEAD) ]; then
  if [[ $RECENT_TAG == v* ]]; then
    VERSION=${RECENT_TAG:1} # Set VERSION to release (with "v") tagged commit
  else
    VERSION=$RECENT_TAG # Set VERSION to non-release (without "v") tagged commit
  fi
else
  VERSION=$(git rev-parse --short HEAD) # Set VERSION to untagged commit
fi
DISTNAME=bitcoin-${VERSION} # Use VERSION in file names
```

Comment (MarcoFalke) [reply contrib/gitian-descriptors/make_git_archive:13]: Why would there be a non-v* tagged commit? This should probably be an assert(false)?

Comment (hebasto) [reply contrib/gitian-descriptors/make_git_archive:13]: It is convenient for making gitian test builds. At least, for me :)

**Comment by MarcoFalke** [comment]
ACK 2aa48edec0101f8a77a2189244fc62722ff7a123

File: contrib/gitian-descriptors/make_git_archive (hunk 1 above)
Comment (MarcoFalke) [reply contrib/gitian-descriptors/make_git_archive:13]: ok, will leave this for other reviewers to bikeshed :)

**Comment by dongcarl** [comment]
Will take a look this week... Testing a few patches on top of this branch RN.

**Comment by dongcarl** [comment]
@hebasto I've opened #18741 based on this! I think that perhaps ab65fe2555eb59633578b4f89c525470e36dd1c3 _could_ be pulled into this PR, but it kind of doesn't make sense without the Guix changes, so I'll leave that decision up to you :relaxed:

**Comment by dongcarl** [comment]
ACK 2aa48edec0101f8a77a2189244fc62722ff7a123
Don't want to delay merging this so let's just do the Guix parts afterwards in #18741

**Comment by adamjonas** [comment]
This should close #16588 if I'm not mistaken.

**Comment by hebasto** [comment]
> This should close #16588 if I'm not mistaken.

The OP has been updated.

**Comment by MarcoFalke** [comment]
@fanquake Anything left to do here?

File: contrib/gitian-descriptors/make_git_archive (hunk 1 above)
Comment (fanquake) [reply contrib/gitian-descriptors/make_git_archive:13]: > It is convenient for making gitian test builds. At least, for me :)

I mentioned on IRC, however I'll just reiterate here, I'd definitely normally NACK unclear/undocumented additions to our build system / release scripts. However given that this should be removed in a follow up PR I think it's ok as is. If you've got patches that are helpful for testing, you can maintain them locally, but please don't just bundle them into a PR along with other changes.

File: test/lint/lint-shell.sh [hunk 2]
```diff
@@ -46,15 +46,17 @@ if ! command -v yq > /dev/null; then
 fi
 
 EXCLUDE_GITIAN=${EXCLUDE}",$(IFS=','; echo "${disabled_gitian[*]}")"
+SHELLCHECK_CMD="shellcheck --external-sources --check-sourced $EXCLUDE_GITIAN"
 for descriptor in $(git ls-files -- 'contrib/gitian-descriptors/*.yml')
 do
-    echo
```
Comment (fanquake) [inline test/lint/lint-shell.sh:51]: I don't know is this is expected or changed here, but running this I see:
```bash
test/lint/lint-shell.sh
2020/04/25 19:36:23 unknown command "contrib/gitian-descriptors/gitian-linux.yml" for "yq"
2020/04/25 19:36:23 unknown command "contrib/gitian-descriptors/gitian-osx-signer.yml" for "yq"
2020/04/25 19:36:23 unknown command "contrib/gitian-descriptors/gitian-osx.yml" for "yq"
2020/04/25 19:36:23 unknown command "contrib/gitian-descriptors/gitian-win-signer.yml" for "yq"
2020/04/25 19:36:23 unknown command "contrib/gitian-descriptors/gitian-win.yml" for "yq"
```

**Review by fanquake — APPROVED** [review]
ACK 2aa48edec0101f8a77a2189244fc62722ff7a123 - I've had a quick look over this, and don't want to block merging if this actually gets as closer to finally having this all sorted out. Obviously we've still got #18741, and after speaking to Carl this morning, there will likely be even more changes after that (not Guix specific).

File: test/lint/lint-shell.sh (hunk 2 above)
Comment (hebasto) [reply test/lint/lint-shell.sh:51]: Could #17361 help? Did you have [`yq`](https://kislyuk.github.io/yq/) installed?

File: contrib/gitian-descriptors/make_git_archive (hunk 1 above)
Comment (hebasto) [reply contrib/gitian-descriptors/make_git_archive:13]: **The rationale for adding that addition:**
Currently, the `gitian-build.py` script allows to make builds from personal repos by passing the `--url` command line argument. Also a branch could be chosen. This change adds an option to point the script to a tagged commit on personal repo as well.

File: test/lint/lint-shell.sh (hunk 2 above)
Comment (fanquake) [reply test/lint/lint-shell.sh:51]: > Did you have yq installed?

Yes. `yq version 3.3.0`

Comment (hebasto) [reply test/lint/lint-shell.sh:51]: > > Did you have yq installed?
> 
> Yes. `yq version 3.3.0`

Not [that](https://github.com/mikefarah/yq/releases) but [this](https://kislyuk.github.io/yq/) one:
```
$ brew install python-yq
```

**Pull request merged**